### ISSUE TITLE
dataclass reported as unhashable when __eq__ is overridden

### ISSUE THREAD

**Issue by atom-andrew** [microsoft/pyright#5928]
**Describe the bug**
frozen dataclasses are reported as unhashable when __eq__ is overridden

Running version 1.1.326

**Code or Screenshots**
This code:

```python
@dataclass(frozen=True)
class Foo:
    x: int

    def __eq__(self, other) -> bool:
        return self.x == other.x


_ = {Foo(x=1)}
```

generates this error:

```
Type "Foo" is not hashable
```

It doesn't matter if we explicitly set `__eq__` or not (or which way).  Per https://docs.python.org/3/library/dataclasses.html,
> If the class already defines [__eq__()](https://docs.python.org/3/reference/datamodel.html#object.__eq__), this parameter is ignored.

**VS Code extension or command-line**
No

**Comment by erictraut** [comment]
Thanks for reporting the issue. The runtime behavior isn't consistent with what I would expect given the behavior documented [in this section](https://docs.python.org/3/library/dataclasses.html#module-contents). I guess it depends on the interpretation of the phrase "the `__hash__()` method of the superclass will be used". I assumed that to mean that no `__hash__` method is explicitly added to the dataclass, but looking at the runtime implementation, it appears that it's copying the `__hash__` method from the superclass to the new dataclass. I'll need to update pyright's logic to reflect the actual runtime behavior.

In the meantime, you can work around this by adding `usafe_hash=True` to the `dataclass` decorator.

**Comment by erictraut** [comment]
This will be addressed in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.327, which I just published. It will also be included in an upcoming release of pylance.
